File: backend/app/utils/permissions.py
"""
Sistema de control de acceso basado en roles para SGT Portal
Define los permisos específicos para cada rol del sistema.
"""
from enum import Enum
from typing import Set, Dict
from app.models.user import UserRole

class Permission(Enum):
    # Permisos básicos - todos los roles
    VIEW_DOCUMENTS = "view_documents"
    VIEW_ORDERS = "view_orders" 
    VIEW_VACATIONS = "view_vacations"
    VIEW_PROFILE = "view_profile"
    EDIT_PROFILE = "edit_profile"
    
    # Permisos de Tráfico
    MANAGE_TRAFFIC = "manage_traffic"
    VIEW_TRAFFIC = "view_traffic"
    EDIT_TRAFFIC = "edit_traffic"
    DELETE_TRAFFIC = "delete_traffic"
    UPLOAD_TRAFFIC_DOCUMENTS = "upload_traffic_documents"
    
    # Permisos de Administración
    MANAGE_USERS = "manage_users"
    VIEW_ALL_USERS = "view_all_users"
    CREATE_USERS = "create_users"
    EDIT_USERS = "edit_users"
    DELETE_USERS = "delete_users"
    RESET_PASSWORDS = "reset_passwords"
    
    MANAGE_SYSTEM = "manage_system"
    VIEW_SYSTEM_SETTINGS = "view_system_settings"
    EDIT_SYSTEM_SETTINGS = "edit_system_settings"
    
    VIEW_ADMIN_DASHBOARD = "view_admin_dashboard"
    VIEW_ALL_DOCUMENTS = "view_all_documents"
    MANAGE_ALL_ORDERS = "manage_all_orders"
    
    # Permisos específicos de documentos
    UPLOAD_DOCUMENTS = "upload_documents"
    DELETE_DOCUMENTS = "delete_documents"
    
    # Permisos de vacaciones
    APPROVE_VACATIONS = "approve_vacations"
    MANAGE_VACATIONS = "manage_vacations"
    
    # Permisos especiales del usuario maestro
    MASTER_ACCESS = "master_access"
    SYSTEM_CONTROL = "system_control"
    UNLIMITED_ACCESS = "unlimited_access"

# Definir permisos por rol
ROLE_PERMISSIONS: Dict[UserRole, Set[Permission]] = {
    UserRole.TRABAJADOR: {
        # Acceso básico: Documentos, Órdenes y Vacaciones
        Permission.VIEW_DOCUMENTS,
        Permission.VIEW_ORDERS,
        Permission.VIEW_VACATIONS,
        Permission.VIEW_PROFILE,
        Permission.EDIT_PROFILE,
        Permission.UPLOAD_DOCUMENTS,
        Permission.DELETE_DOCUMENTS,  # Solo sus propios documentos
    },
    
    UserRole.P_TALLER: {
        # Personal de Taller - acceso a documentos, vacaciones, tráfico y viajes
        Permission.VIEW_DOCUMENTS,
        Permission.VIEW_ORDERS,       # Acceso a viajes
        Permission.VIEW_VACATIONS,
        Permission.VIEW_PROFILE,
        Permission.EDIT_PROFILE,
        Permission.UPLOAD_DOCUMENTS,
        Permission.DELETE_DOCUMENTS,  # Solo sus propios documentos
        Permission.VIEW_TRAFFIC,  # Acceso a carpeta de tráfico (solo lectura)
    },
    
    UserRole.TRAFICO: {
        # Todo lo de TRABAJADOR + gestión completa de tráfico
        Permission.VIEW_DOCUMENTS,
        Permission.VIEW_ORDERS,
        Permission.VIEW_VACATIONS,
        Permission.VIEW_PROFILE,
        Permission.EDIT_PROFILE,
        Permission.UPLOAD_DOCUMENTS,
        Permission.DELETE_DOCUMENTS,
        
        # Permisos específicos de tráfico
        Permission.MANAGE_TRAFFIC,
        Permission.VIEW_TRAFFIC,
        Permission.EDIT_TRAFFIC,
        Permission.DELETE_TRAFFIC,
        Permission.UPLOAD_TRAFFIC_DOCUMENTS,
    },
    
    # Nuevo rol: ADMINISTRACION (backoffice limitado)
    # Puede acceder a módulos de gestor docs, vacaciones, dietas (solo lectura si el endpoint lo soporta),
    # tráfico (solo ver), y viajes. No tiene permisos de gestión global ni de usuarios.
    UserRole.ADMINISTRACION: {
        Permission.VIEW_DOCUMENTS,
        Permission.VIEW_ORDERS,
        Permission.VIEW_VACATIONS,
        Permission.VIEW_PROFILE,
        Permission.EDIT_PROFILE,
        Permission.UPLOAD_DOCUMENTS,
        # tráfico solo ver
        Permission.VIEW_TRAFFIC,
        # viajes (módulo orders/trips en este contexto)
        # Nota: la capacidad de ver todos los viajes está restringida a ADMINISTRADOR/MASTER_ADMIN en los endpoints
    },
    
    UserRole.ADMINISTRADOR: {
        # Acceso completo a todo el sistema
        Permission.VIEW_DOCUMENTS,
        Permission.VIEW_ORDERS,
        Permission.VIEW_VACATIONS,
        Permission.VIEW_PROFILE,
        Permission.EDIT_PROFILE,
        Permission.UPLOAD_DOCUMENTS,
        Permission.DELETE_DOCUMENTS,
        
        Permission.MANAGE_TRAFFIC,
        Permission.VIEW_TRAFFIC,
        Permission.EDIT_TRAFFIC,
        Permission.DELETE_TRAFFIC,
        Permission.UPLOAD_TRAFFIC_DOCUMENTS,
        
        Permission.MANAGE_USERS,
        Permission.VIEW_ALL_USERS,
        Permission.CREATE_USERS,
        Permission.EDIT_USERS,
        Permission.DELETE_USERS,
        Permission.RESET_PASSWORDS,
        
        Permission.MANAGE_SYSTEM,
        Permission.VIEW_SYSTEM_SETTINGS,
        Permission.EDIT_SYSTEM_SETTINGS,
        
        Permission.VIEW_ADMIN_DASHBOARD,
        Permission.VIEW_ALL_DOCUMENTS,
        Permission.MANAGE_ALL_ORDERS,
        
        Permission.APPROVE_VACATIONS,
        Permission.MANAGE_VACATIONS,
    },
    
    UserRole.MASTER_ADMIN: {
        # Acceso absolutamente ilimitado - todos los permisos existentes y futuros
        Permission.VIEW_DOCUMENTS,
        Permission.VIEW_ORDERS,
        Permission.VIEW_VACATIONS,
        Permission.VIEW_PROFILE,
        Permission.EDIT_PROFILE,
        Permission.UPLOAD_DOCUMENTS,
        Permission.DELETE_DOCUMENTS,
        
        Permission.MANAGE_TRAFFIC,
        Permission.VIEW_TRAFFIC,
        Permission.EDIT_TRAFFIC,
        Permission.DELETE_TRAFFIC,
        Permission.UPLOAD_TRAFFIC_DOCUMENTS,
        
        Permission.MANAGE_USERS,
        Permission.VIEW_ALL_USERS,
        Permission.CREATE_USERS,
        Permission.EDIT_USERS,
        Permission.DELETE_USERS,
        Permission.RESET_PASSWORDS,
        
        Permission.MANAGE_SYSTEM,
        Permission.VIEW_SYSTEM_SETTINGS,
        Permission.EDIT_SYSTEM_SETTINGS,
        
        Permission.VIEW_ADMIN_DASHBOARD,
        Permission.VIEW_ALL_DOCUMENTS,
        Permission.MANAGE_ALL_ORDERS,
        
        Permission.APPROVE_VACATIONS,
        Permission.MANAGE_VACATIONS,
        
        # Permisos especiales exclusivos del usuario maestro
        Permission.MASTER_ACCESS,
        Permission.SYSTEM_CONTROL,
        Permission.UNLIMITED_ACCESS,
    }
}

def has_permission(user_role: UserRole, permission: Permission) -> bool:
    """
    Verifica si un rol tiene un permiso específico.
    El usuario maestro siempre tiene todos los permisos.
    
    Args:
        user_role: Rol del usuario
        permission: Permiso a verificar
        
    Returns:
        bool: True si el rol tiene el permiso, False en caso contrario
    """
    # El usuario maestro siempre tiene acceso a todo
    if user_role == UserRole.MASTER_ADMIN:
        return True
        
    return permission in ROLE_PERMISSIONS.get(user_role, set())

def is_master_admin(user) -> bool:
    """
    Verifica si un usuario es el administrador maestro.
    
    Args:
        user: Usuario a verificar (puede ser User de DB o MasterAdminUser)
        
    Returns:
        bool: True si es el usuario maestro
    """
    return hasattr(user, 'role') and user.role == UserRole.MASTER_ADMIN

def get_user_permissions(user_role: UserRole) -> Set[Permission]:
    """
    Obtiene todos los permisos de un rol específico.
    
    Args:
        user_role: Rol del usuario
        
    Returns:
        Set[Permission]: Conjunto de permisos del rol
    """
    return ROLE_PERMISSIONS.get(user_role, set())

def can_access_module(user_role: UserRole, module: str) -> bool:
    """
    Verifica si un rol puede acceder a un módulo específico.
    El usuario maestro siempre puede acceder a todos los módulos.
    
    Args:
        user_role: Rol del usuario
        module: Nombre del módulo (dashboard, traffic, documents, etc.)
        
    Returns:
        bool: True si puede acceder, False en caso contrario
    """
    # El usuario maestro siempre tiene acceso a todo
    if user_role == UserRole.MASTER_ADMIN:
        return True
        
    module_permissions = {
        'documents': Permission.VIEW_DOCUMENTS,
        'orders': Permission.VIEW_ORDERS,
        'vacations': Permission.VIEW_VACATIONS,
        'traffic': Permission.VIEW_TRAFFIC,
        'users': Permission.VIEW_ALL_USERS,
        'settings': Permission.VIEW_SYSTEM_SETTINGS,
        'gestor': Permission.VIEW_ADMIN_DASHBOARD,
    }
    
    required_permission = module_permissions.get(module)
    if not required_permission:
        return False
        
    return has_permission(user_role, required_permission)

# Decorator para verificar permisos en endpoints
from functools import wraps
from fastapi import HTTPException, status
from app.models.user import User

def require_permission(permission: Permission):
    """
    Decorator para verificar permisos en endpoints de FastAPI.
    El usuario maestro siempre tiene todos los permisos.
    
    Args:
        permission: Permiso requerido
    """
    def decorator(func):
        @wraps(func)
        async def wrapper(*args, **kwargs):
            # Buscar el usuario actual en los argumentos
            current_user = None
            for arg in args:
                if hasattr(arg, 'role') and hasattr(arg, 'dni_nie'):
                    current_user = arg
                    break
            
            # Buscar en kwargs
            if not current_user:
                current_user = kwargs.get('current_user')
            
            if not current_user:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Usuario no autenticado"
                )
            
            # Obtener el rol del usuario
            user_role = getattr(current_user, 'role', None)
            if not user_role:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Usuario sin rol asignado"
                )
            
            if not has_permission(user_role, permission):
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail="No tienes permisos para realizar esta acción"
                )
            
            return await func(*args, **kwargs)
        return wrapper
    return decorator

def require_role(*allowed_roles: UserRole):
    """
    Decorator para verificar roles específicos en endpoints.
    El usuario maestro siempre tiene acceso.
    
    Args:
        allowed_roles: Roles permitidos
    """
    def decorator(func):
        @wraps(func)
        async def wrapper(*args, **kwargs):
            current_user = None
            for arg in args:
                if hasattr(arg, 'role') and hasattr(arg, 'dni_nie'):
                    current_user = arg
                    break
            
            if not current_user:
                current_user = kwargs.get('current_user')
            
            if not current_user:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Usuario no autenticado"
                )
            
            # Obtener el rol del usuario
            user_role = getattr(current_user, 'role', None)
            if not user_role:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Usuario sin rol asignado"
                )
            
            # El usuario maestro siempre tiene acceso
            if user_role == UserRole.MASTER_ADMIN:
                return await func(*args, **kwargs)
            
            if user_role not in allowed_roles:
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail="No tienes el rol necesario para acceder a este recurso"
                )
            
            return await func(*args, **kwargs)
        return wrapper
    return decorator

File: frontend/src/utils/permissions.ts
import type { User } from '../types';

// Tipos de permisos
export enum Permission {
  // Módulo Dashboard
  VIEW_DASHBOARD = 'view_dashboard',
  
  // Módulo Documentos
  VIEW_DOCUMENTS = 'view_documents',
  UPLOAD_DOCUMENTS = 'upload_documents',
  MANAGE_DOCUMENTS = 'manage_documents',
  
  // Módulo Órdenes
  VIEW_TRIPS = 'view_trips',
  MANAGE_TRIPS = 'manage_trips',
  
  // Módulo Vacaciones
  VIEW_VACATIONS = 'view_vacations',
  MANAGE_VACATIONS = 'manage_vacations',

  // Módulo Dietas (solo administradores)
  VIEW_DIETAS = 'view_dietas',
  
  // Módulo Tráfico
  VIEW_TRAFFIC = 'view_traffic',
  MANAGE_TRAFFIC = 'manage_traffic',
  
  // Módulo Usuarios
  VIEW_USERS = 'view_users',
  MANAGE_USERS = 'manage_users',
  
  // Panel de Documentación
  VIEW_GESTOR = 'view_gestor',
  
  // Subida Masiva (solo administradores)
  MASS_UPLOAD = 'mass_upload',
  
  // Configuración
  VIEW_SETTINGS = 'view_settings',
  MANAGE_SETTINGS = 'manage_settings',
}

// Permisos por rol
const ROLE_PERMISSIONS: Record<string, Permission[]> = {
  MASTER_ADMIN: [
    // Usuario maestro tiene TODOS los permisos
    Permission.VIEW_DASHBOARD,
    Permission.VIEW_DOCUMENTS,
    Permission.UPLOAD_DOCUMENTS,
    Permission.MANAGE_DOCUMENTS,
  Permission.VIEW_TRIPS,
  Permission.MANAGE_TRIPS,
    Permission.VIEW_VACATIONS,
    Permission.MANAGE_VACATIONS,
  Permission.VIEW_DIETAS,
    Permission.VIEW_TRAFFIC,
    Permission.MANAGE_TRAFFIC,
    Permission.VIEW_USERS,
    Permission.MANAGE_USERS,
    Permission.VIEW_GESTOR,
    Permission.MASS_UPLOAD,
    Permission.VIEW_SETTINGS,
    Permission.MANAGE_SETTINGS,
  ],
  TRABAJADOR: [
    Permission.VIEW_DOCUMENTS,
    Permission.UPLOAD_DOCUMENTS,
  Permission.VIEW_TRIPS, // Conductores ven viajes
    Permission.VIEW_VACATIONS,
  ],
  P_TALLER: [
    // Personal de Taller - acceso a documentos, vacaciones, tráfico y viajes
    Permission.VIEW_DOCUMENTS,
    Permission.UPLOAD_DOCUMENTS,
    Permission.VIEW_TRIPS,        // Acceso a viajes
    Permission.MANAGE_TRIPS,      // Puede crear y gestionar viajes
    Permission.VIEW_VACATIONS,
    Permission.VIEW_TRAFFIC,      // Acceso a carpeta de tráfico
  ],
  TRAFICO: [
    Permission.VIEW_DOCUMENTS,
    Permission.UPLOAD_DOCUMENTS,
    // Permission.VIEW_ORDERS - REMOVIDO: Tráfico no ve órdenes
    Permission.VIEW_VACATIONS,
    Permission.VIEW_TRAFFIC,
    Permission.MANAGE_TRAFFIC,
  Permission.VIEW_DASHBOARD, // añadido para permitir acceso al dashboard
  ],
  ADMINISTRADOR: [
    Permission.VIEW_DASHBOARD,
    Permission.VIEW_DOCUMENTS,
    Permission.UPLOAD_DOCUMENTS,
    Permission.MANAGE_DOCUMENTS,
  // Órdenes habilitado nuevamente para administradores:
  Permission.VIEW_TRIPS,
  Permission.MANAGE_TRIPS,
    Permission.VIEW_VACATIONS,
    Permission.MANAGE_VACATIONS,
  Permission.VIEW_DIETAS,
    Permission.VIEW_TRAFFIC,
    Permission.MANAGE_TRAFFIC,
    Permission.VIEW_USERS,
    Permission.MANAGE_USERS,
    Permission.VIEW_GESTOR,
    Permission.MASS_UPLOAD,
    // Permission.VIEW_SETTINGS - REMOVIDO: Solo MASTER_ADMIN puede ver configuración
    // Permission.MANAGE_SETTINGS - REMOVIDO: Solo MASTER_ADMIN puede gestionar configuración
  ],
  ADMINISTRACION: [
    // Puede ver Documentos (propios), Viajes, Vacaciones, Dietas (vista), Tráfico (solo ver)
    Permission.VIEW_DOCUMENTS,
    Permission.UPLOAD_DOCUMENTS, // si aplica a sus propios documentos
    Permission.VIEW_TRIPS,
    Permission.VIEW_VACATIONS,
    Permission.VIEW_DIETAS,
    Permission.VIEW_TRAFFIC,
  ],
};

// Función para verificar si un usuario tiene un permiso específico
export const hasPermission = (user: User | null, permission: Permission): boolean => {
  if (!user || !user.role) return false;
  
  // El usuario maestro tiene todos los permisos, siempre
  if (user.role === 'MASTER_ADMIN') return true;
  
  const userPermissions = ROLE_PERMISSIONS[user.role] || [];
  return userPermissions.includes(permission);
};

// Función para verificar si un usuario tiene alguno de varios permisos
export const hasAnyPermission = (user: User | null, permissions: Permission[]): boolean => {
  return permissions.some(permission => hasPermission(user, permission));
};

// Función para obtener todos los permisos de un usuario
export const getUserPermissions = (user: User | null): Permission[] => {
  if (!user || !user.role) return [];
  return ROLE_PERMISSIONS[user.role] || [];
};

// Función para verificar si un usuario puede acceder a una ruta específica
export const canAccessRoute = (user: User | null, route: string): boolean => {
  if (!user) return false;
  
  switch (route) {
    case '/':
      return hasPermission(user, Permission.VIEW_DOCUMENTS);
    case '/dashboard':
      return hasPermission(user, Permission.VIEW_DASHBOARD);
    case '/mass-upload':
      return hasPermission(user, Permission.MASS_UPLOAD);
    case '/documents':
      return hasPermission(user, Permission.VIEW_DOCUMENTS);
    case '/trips':
      return hasPermission(user, Permission.VIEW_TRIPS);
    case '/vacations':
      return hasPermission(user, Permission.VIEW_VACATIONS);
    case '/traffic':
      return hasPermission(user, Permission.VIEW_TRAFFIC);
    case '/users':
      return hasPermission(user, Permission.VIEW_USERS);
    case '/gestor':
      return hasPermission(user, Permission.VIEW_GESTOR);
    case '/dietas':
      return hasPermission(user, Permission.VIEW_DIETAS);
    case '/recursos':
      // Página informativa accesible a quienes pueden ver documentos (trabajadores y superiores)
      return hasPermission(user, Permission.VIEW_DOCUMENTS);
    case '/settings':
      // Solo el usuario maestro puede acceder a configuración
      return user.role === 'MASTER_ADMIN';
  // case '/profile': eliminado
    default:
      return false;
  }
};

// Función para obtener el texto de un rol en español
export const getRoleText = (role: string): string => {
  switch (role) {
    case 'MASTER_ADMIN':
      return 'Admin Master';
    case 'ADMINISTRADOR':
      return 'Administrador';
    case 'ADMINISTRACION':
      return 'Administración';
    case 'TRAFICO':
      return 'Tráfico';
    case 'TRABAJADOR':
      return 'Trabajador';
    case 'P_TALLER':
      return 'P. Taller';
    default:
      return 'Usuario';
  }
};

// Función para obtener la primera ruta permitida para un usuario
export const getDefaultRoute = (user: User | null): string => {
  if (!user) return '/login';
  
  // Verificar rutas en orden de prioridad
  if (hasPermission(user, Permission.VIEW_DASHBOARD)) return '/dashboard';
  if (hasPermission(user, Permission.VIEW_DOCUMENTS)) return '/';
  if (hasPermission(user, Permission.VIEW_TRIPS)) return '/trips';
  if (hasPermission(user, Permission.VIEW_VACATIONS)) return '/vacations';
  if (hasPermission(user, Permission.VIEW_TRAFFIC)) return '/traffic';
  if (hasPermission(user, Permission.VIEW_USERS)) return '/users';
  if (hasPermission(user, Permission.VIEW_GESTOR)) return '/gestor';
  
  return '/login'; // Fallback actualizado (perfil eliminado)
};

// Función para obtener el color de un rol
export const getRoleColor = (role: string): string => {
  switch (role) {
    case 'MASTER_ADMIN':
      return '#9c27b0'; // Púrpura para el usuario maestro
    case 'ADMINISTRADOR':
      return '#d32f2f'; // Rojo
    case 'ADMINISTRACION':
      return '#d4a574'; // Dorado suave para administración
    case 'TRAFICO':
      return '#1976d2'; // Azul
    case 'TRABAJADOR':
      return '#388e3c'; // Verde
    case 'P_TALLER':
      return '#ff9800'; // Naranja para personal de taller
    default:
      return '#666666'; // Gris
  }
};
